cpp/src/arrow/compute/kernels/scalar_nested.cc

```diff
@@ -271,11 +318,30 @@ struct StructFieldFunctor {
 
 Result<TypeHolder> ResolveStructFieldType(KernelContext* ctx,
                                           const std::vector<TypeHolder>& types) {
-  const auto& options = OptionsWrapper<StructFieldOptions>::Get(ctx);
+  const auto& field_ref = OptionsWrapper<StructFieldOptions>::Get(ctx).field_ref;
   const DataType* type = types.front().type;
-  for (const auto& index : options.indices) {
-    RETURN_NOT_OK(StructFieldFunctor::CheckIndex(index, *type));
-    type = type->field(index)->type().get();
+
+  if (field_ref.IsName()) {
+    for (const auto& path : field_ref.FindAll(*type)) {
+      for (const auto& index : path.indices()) {
+        RETURN_NOT_OK(StructFieldFunctor::CheckIndex(index, *type));
+        type = type->field(index)->type().get();
+      }
+    }
+  } else {
+    DCHECK(field_ref.IsFieldPath() || field_ref.IsNested());
+
+    FieldPath field_path;
+    if (field_ref.IsNested()) {
+      ARROW_ASSIGN_OR_RAISE(field_path, field_ref.FindOne(*type));
+    } else {
+      field_path = *field_ref.field_path();
+    }
```

Comment:
I am wondering if it is actually needed to do this if/else. Also a FieldRef backed by a field path has a FindOne method, and in that case it will basically return the same field path, while still checking that it is a valid field path (not out of bounds)

Reply:
Seems like it is needed, removing if/else and using `field_ref.FindOne(*type)` will abort when `pc.struct_field(arr, [])` (empty indices) with:
```
pyarrow/tests/test_compute.py::test_struct_fields_options /home/milesg/Projects/arrow/cpp/src/arrow/type.cc:1360:  Check failed: _s.ok() Operation failed: maybe_field.status()
Bad status: Invalid: empty indices cannot be traversed
```
Seems like another if statement would be needed to handle that case, so maybe stick with this?

Reply:
I am wondering if it is actually intentional that we support empty index (I don't see it covered in the C++ tests?). That doesn't sound very useful? (it just returns the original array?)

Reply:
The empty index case was in the Python tests. 

I suppose it may be useful if only to serve the purpose of idempotent behavior, similar to `[:]` or `start==stop` -> empty lists in the `list_slice` kernel. Albeit those are unique to slicing.